kafkaesk/ext/logging/formatter.py

```diff
@@ -0,0 +1,81 @@
+from .record import factory
+from .record import PydanticLogRecord
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Tuple
+from typing import Type
+
+import logging
+import pydantic
+
+
+class BaseLogFormat(pydantic.BaseModel):
+    timestamp: str = pydantic.Field(alias="asctime")
+    logger: str = pydantic.Field(alias="name")
+    severity: str = pydantic.Field(alias="levelname")
+    level: int = pydantic.Field(alias="levelno")
+    message: str
+    exception: Optional[str] = pydantic.Field(alias="exc_type")
+    trace: Optional[str] = pydantic.Field(alias="exc_text")
+    stack: Optional[str] = pydantic.Field(alias="stack_info")
+
+
+class PydanticLogModel(pydantic.BaseModel):
```

Comment:
Seems that this is the main Pydantic class that is gonna be exposed, amb I right? which will be the one used for later on  for subscribing to messages, as it is showed here [1].

Could we make this class more explicit in terms of fields that this class will have? I'm wondering why not add the fields exposed by the `BaseLogFormat`? and, what could we do for making the extras even more visible?

Thoughts?


[1] https://github.com/onna/kafkaesk/pull/12/files#diff-9fd9cabda898f0c595cc27aba631f384R35

Reply:
Been thinking about this and I'm not sure the right answer.  I could use some feedback on my thoughts here.

We have three different types of data that are represented by Pydantic classes here:
1. `BaseLogFormat` - This is roughly equivalent to a `fmt` string passed to vanilla python formatters.  This extracts information from a log record and formats the output message.  While a default Pydantic class is provided with the library this may be replaced by a custom class by the user.
2. `LogModel` - These classes are passed in as args to a log call and are completely dynamic.  We have no knowledge of what these might contain (or if any will be present) before the log call is made.
3. `PydanticLogModel` - This is basically a wrapper model that gets populated by the `BaseLogFormat` and any `LogModel` classes that were passed to the log call.  The only real reason this exists is so that the data we pass to Kafkaesk is in the correct format.  I'm hesitant to add the `BaseLogFormat` fields to this because `BaseLogFormat` may not be the class that's actually used by the formatter.

My first implementation was a bit more restrictive; when creating a `PydanticKafkaeskHandler` instance you had to provide a Pydantic class.  That class was used to parse all information and and instance of the class is what was passed to Kafkaesk.  This ended up causing several issues:

1. It introduced a trade off between a lot of boilerplate for every log call made or not having Mypy checking when the logging call was made.
2. Every possible attribute that an application may write to a log needed to be present on the logger or many loggers needed to be created to handle different schemas.

For now I'm thinking that not populating the body of `PydanticLogModel` is going to be the best option.  It's not strictly `correct` to use a Pydantic class in this manner and doesn't entirely fit with the Kafkaesk way of doing thing but messages produced by the logger aren't intended to be generally consumed.  We will likely have a dedicated listener (either Confluent Connect's Elasticsearch Sink application or a custom consumer) that doesn't actually care about the schema, it will just pull out the raw data and forward it to Elasticsearch.

Reply:
> For now I'm thinking that not populating the body of PydanticLogModel is going to be the best option. It's not strictly correct to use a Pydantic class in this manner and doesn't entirely fit with the Kafkaesk way of doing thing but messages produced by the logger aren't intended to be generally consumed. We will likely have a dedicated listener (either Confluent Connect's Elasticsearch Sink application or a custom consumer) that doesn't actually care about the schema, it will just pull out the raw data and forward it to Elasticsearch.

You made a very valid good point here, let's see how it works on the consuming side!